### ISSUE TITLE
io.js support

### ISSUE THREAD

**Issue by appsforartists** [sass/node-sass#627]
Presently, trying to run `sass-loader` against the latest `node-sass@2.0.0-beta` (the latest in npm) on `io.js@1.0.1` appears to succeed, but upon execution yields a `Module did not self-register` error.  If you follow the advice of #563 and remove the binding altogether, you'll get this instead:

```
Module build failed: TypeError: Cannot read property 'render' of undefined
    at Object.module.exports.render (node_modules/node-sass/lib/index.js:230:26)
    at Object.module.exports (node_modules/sass-loader/index.js:53:10)
```

**Comment by am11** [comment]
Please read the comment on the original issue: https://github.com/iojs/io.js/issues/373#issuecomment-69883613, we are currently supporting latest stable version of node.js. 

io.js runtime support is not on our TODO list yet.

**Comment by snostorm** [comment]
On https://github.com/sass/node-sass/issues/563#issuecomment-70072413 I've hinted at the code changes needed to help make io.js work incase anybody has a chance to put together a PR. (I won't have time today for sure, although if pinged I can pass along more details on my local diff -- which wouldn't yet work for 0.10 users.)

**Comment by mnutt** [comment]
https://github.com/iojs/io.js/issues/456

**Comment by am11** [comment]
This is addressed by 452c698644014265336659da65aa0b7a2d22e6ed and 0887f836e0f68199e2f16368296d4bb80370caaf via #630.

Credit: @mnutt, @snostorm.

**Comment by am11** [comment]
Just to be clear: with https://github.com/sass/node-sass/pull/630, it is *possible* to build the binary against io.js runtime. The one that `npm install node-sass` downloads, only works with node v0.10.x.

I am not sure if the binary built by one version of a runtime should work seamlessly on any other runtime-version. Or are we expected to make a separate set of binaries for io.js? If so, then it will not be available with <kbd>v2</kbd>. Meanwhile you would need to rebuild the binary manually for io.js.

Similarly, even the binary that is built with node v0.10.35, does not work with v0.11.14. Again, you would need to rebuild it manually. 

Note that this is a separate issue than making the Linux binaries cross-distro compatible: https://github.com/sass/node-sass/issues/602 and https://github.com/sass/node-sass/issues/517. Compatibility is a biggest challenge for us, as our upstream libsass is using C99 and C++11 STL features, our users require it to work on old CentOS 5.5+ with newer libc & older libstdc++ (compared to debian counterparts) and the minimum gcc & g++ compiler version required to compile our binding is v4.6.

Now you know how io.js would make the binary building and distribution process complicated, especially for node-sass.

**Comment by kkoopa** [comment]
A module must be rebuilt whenever `NODE_MODULE_VERSION` changes. It changed from joyent/node 0.10 -> joyent/node 0.11, from joyent/node 0.11 -> iojs/io.js 1.0, and atom/atom-shell has its own fork going with yet another `NODE_MODULE_VERSION`.

**Comment by appsforartists** [comment]
iojs support has been merged for days, but the version hasn't been bumped yet.  Can we please release to `npm` so we can test this out?

**Comment by am11** [comment]
@appsforartists see https://github.com/sass/node-sass/issues/627#issuecomment-70394304

**Comment by appsforartists** [comment]
I must have misunderstood.  I thought adding iojs binaries to the npm module what [@browniefed was doing](https://github.com/browniefed/node-sass-binaries).

So to use this in iojs, would you just need to follow [these steps](https://github.com/sass/node-sass#rebuilding-binaries)?

**Comment by am11** [comment]
Pretty much the same. Note that the official node-gyp package is yet to receive iojs love: https://github.com/TooTallNate/node-gyp/pull/564. Meanwhile, we can use the one that is bundled with iojs as @mnutt [mentioned here](https://github.com/sass/node-sass/pull/630#issue-54546074).

Sometime ago, we were thinking about expanding the platforms support in terms of binaries (FreeBSD, Solaris). But support for ioj.s and kind is certainly another important dimension. The show stopper at the moment is compatibility for Linux binaries. If we release without caring about older versions of linux platform, it will make masses unhappy (that is why I don't want to rush but rather work out the permanent solution this time for this recurring problem). Therefore, it is quite hard to say when v2.0.0 will be released and this expansion plan will move forward.

**Comment by xzyfer** [comment]
FYI http://docs.travis-ci.com/user/build-environment-updates/2015-02-03/#iojs_support

**Comment by AntouanK** [comment]
A bit confused...
Should we be able to build node-sass on `iojs` or not?

That's my output trying to build node-sass ( after cloning the repo, etc )

``` shell
➜  node-sass git:(master) node --version && npm --version && node-gyp --version
v1.0.4
2.3.0
v1.0.2
➜  node-sass git:(master) sudo node-gyp rebuild
gyp info it worked if it ends with ok
gyp info using node-gyp@1.0.2
gyp info using node@1.0.4 | darwin | x64
child_process: customFds option is deprecated, use stdio instead.
gyp info spawn python
gyp info spawn args [ '/usr/local/lib/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/antouank/_REPOS_/test/client/node_modules/grunt-sass/node_modules/node-sass/build/config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/usr/local/lib/node_modules/node-gyp/addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/antouank/.node-gyp/1.0.4/common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=/Users/antouank/.node-gyp/1.0.4',
gyp info spawn args   '-Dmodule_root_dir=/Users/antouank/_REPOS_/test/client/node_modules/grunt-sass/node_modules/node-sass',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'build',
gyp info spawn args   '-Goutput_dir=.' ]
xcode-select: error: tool 'xcodebuild' requires Xcode, but active developer directory '/Library/Developer/CommandLineTools' is a command line tools instance

xcode-select: error: tool 'xcodebuild' requires Xcode, but active developer directory '/Library/Developer/CommandLineTools' is a command line tools instance

gyp info spawn make
gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
  CXX(target) Release/obj.target/binding/src/binding.o
In file included from ../src/binding.cpp:1:
../node_modules/nan/nan.h:1848:10: error: no matching function for call to 'Encode'
  return node::Encode(
         ^~~~~~~~~~~~
/Users/antouank/.node-gyp/1.0.4/src/node.h:251:34: note: candidate function not viable: cannot convert argument of incomplete type 'const void *' to 'const char *'
NODE_EXTERN v8::Local<v8::Value> Encode(v8::Isolate* isolate,
                                 ^
/Users/antouank/.node-gyp/1.0.4/src/node.h:257:34: note: candidate function not viable: requires 3 arguments, but 4 were provided
NODE_EXTERN v8::Local<v8::Value> Encode(v8::Isolate* isolate,
                                 ^
/Users/antouank/.node-gyp/1.0.4/src/node.h:262:45: note: candidate function not viable: requires at most 3 arguments, but 4 were provided
                inline v8::Local<v8::Value> Encode(
                                            ^
/Users/antouank/.node-gyp/1.0.4/src/node.h:43:53: note: expanded from macro 'NODE_DEPRECATED'
#define NODE_DEPRECATED(msg, fn) V8_DEPRECATED(msg, fn)
                                                    ^
/Users/antouank/.node-gyp/1.0.4/deps/v8/include/v8config.h:332:45: note: expanded from macro 'V8_DEPRECATED'
# define V8_DEPRECATED(message, declarator) declarator
                                            ^
1 error generated.
make: *** [Release/obj.target/binding/src/binding.o] Error 1
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/node-gyp/lib/build.js:267:23)
gyp ERR! stack     at ChildProcess.emit (events.js:101:17)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:1038:12)
gyp ERR! System Darwin 14.0.0
gyp ERR! command "node" "/usr/local/bin/node-gyp" "rebuild"
gyp ERR! cwd /Users/antouank/_REPOS_/test/client/node_modules/grunt-sass/node_modules/node-sass
gyp ERR! node -v v1.0.4
gyp ERR! node-gyp -v v1.0.2
gyp ERR! not ok
➜  node-sass git:(master)
```

**Comment by am11** [comment]
@xzyfer, great thanks! Also http://www.appveyor.com/docs/lang/nodejs-iojs. We will definitely incorporate io.js in our CI matrices. :)

@AntouanK 
I have [updated the dependencies](https://github.com/am11/node-sass/commit/196464492b6fcbda8f161c428ce35b786b9d8197) and able to reproduce this issue.

@kkoopa, here is the first error:

```powershell
C:\Users\adeel\Source\Repos\node-sass\node_modules\nan\nan.h(1851):
error C2665: 'node::Encode' : none of the 3 overloads could convert all the
 argument types (..\src\binding.cpp) [C:\Users\adeel\Source\Repos\node-sass
\build\binding.vcxproj]
```

Full log: https://gist.github.com/am11/68a2022bbe49b193dcd0.

I am using [pangyp](https://github.com/rvagg/pangyp). Following are the steps to reproduce the bug:

```bash
git clone https://github.com/sass/node-sass --recursive; cd node-sass
git submodule update --recursive --init
npm install; npm install pangyp
iojs node_modules/pangyp/bin/node-gyp rebuild
```

Is it a known bug or am I doing it wrong? :)

**Comment by AntouanK** [comment]
@am11 Sorry, my english might not be that good. After updating the dependencies, are you able to reproduce the errors I have or not?

**Comment by am11** [comment]
@AntouanK, yes I am able to reproduce the issue you have reported. The error is coming from [nan](https://github.com/rvagg/nan) code.

**Comment by AntouanK** [comment]
@am11 Ok. Let us know if you want to use any other log from the error.

**Comment by Globegitter** [comment]
I just tried the very latest master with iojs 1.1 on Mac OSX 64 and not getting past `npm install`.

```
git clone https://github.com/sass/node-sass --recursive; cd node-sass
git submodule update --recursive --init
npm install
```
leads to:
```
module.js:341
  Module._extensions[extension](this, filename);
                               ^
Error: Module did not self-register.
    at Error (native)
    at Module.load (module.js:341:32)
    at Function.Module._load (module.js:296:12)
    at Module.require (module.js:351:17)
    at require (module.js:370:17)
    at Object.<anonymous> (node-sass/lib/index.js:198:15)
    at Module._compile (module.js:446:26)
    at Object.Module._extensions..js (module.js:464:10)
    at Module.load (module.js:341:32)
    at Function.Module._load (module.js:296:12)
```

And within the `npm-debug.log` I get the following:

```
37 info node-sass@2.0.0-beta Failed to exec postinstall script
38 verbose stack Error: node-sass@2.0.0-beta postinstall: `node scripts/build.js`
38 verbose stack Exit status 1
38 verbose stack     at EventEmitter.<anonymous> (/usr/local/lib/node_modules/npm/lib/utils/lifecycle.js:213:16)
38 verbose stack     at EventEmitter.emit (events.js:100:17)
38 verbose stack     at ChildProcess.<anonymous> (/usr/local/lib/node_modules/npm/lib/utils/spawn.js:14:12)
38 verbose stack     at ChildProcess.emit (events.js:100:17)
38 verbose stack     at maybeClose (child_process.js:981:16)
38 verbose stack     at Process.ChildProcess._handle.onexit (child_process.js:1053:5)
39 verbose pkgid node-sass@2.0.0-beta
40 verbose cwd /Users/markus/Projects/Git/node-sass
41 error Darwin 14.1.0
42 error argv "/usr/local/bin/iojs" "/usr/local/bin/npm" "install"
43 error node v1.1.0
44 error npm  v2.4.1
45 error code ELIFECYCLE
46 error node-sass@2.0.0-beta postinstall: `node scripts/build.js`
46 error Exit status 1
```

Also when commenting out `var binding = require(getBinding());` which I read somewhere in one of the issues here I am getting another error on `npm install`. `pangyp rebuild` does not seem to work for me either.

**Comment by Xinkai** [comment]
Yesterday I built node-sass with io.js 1.1 on a Linux machine on top of  98f300e45211a4a87a422dc19a2ac4b7c340524e. I used pangyp, and the only change that I needed to make was to change the the version of `nan` to `^1.5.3` in `package.json` file.

The Encode error mentioned by @AntouanK was indeed a nan bug: rvagg/nan#267

**Comment by Globegitter** [comment]
@Xinkai How exactly did you get to build? I tried fresh and still getting the `Module did not self-register.` Error on `npm install` until I comment out a few lines in the `lib/index.js`.

**Comment by Xinkai** [comment]
Follow instructions at https://github.com/sass/node-sass#rebuilding-binaries, except replace `node-gyp` to `pangyp`.

**Comment by Globegitter** [comment]
So @Xinkai are you getting an error on `npm install` as well and just ignoring it?

**Comment by Xinkai** [comment]
I do recall having problems with `npm install`. Something about testing an existing binary. I remember I renamed that binary under `vendor`. Not sure if that's relevant.

Can you try that? I don't remember clearly because I tried multiple things.

**Comment by Globegitter** [comment]
I actually got it to work completely now, even when installing as dependency: https://github.com/Globegitter/node-sass

It is a quite hacky workaround, but still:
I updated `nan` to use https://github.com/rvagg/nan/pull/265.
I changed `https://github.com/Globegitter/node-sass/blob/master/scripts/build.js` to use `pangyp` and to always run `return build(options);` which should now lead to a successful build immediately.

To get it to work as an npm package, (`"node-sass": "Globegitter/node-sass"`) I had to add a git clone preinstall script: https://github.com/Globegitter/node-sass/blob/master/package.json#L24

Hope this helps someone as a temporary workaround :)

**Comment by Xinkai** [comment]
I recollected my memory and tested it out. Hope it helps someone else.

``` bash
npm install -g pangyp

git clone --recursive https://github.com/sass/node-sass.git
cd node-sass
git submodule update --init --recursive

# Edit package.json file: change nan to ^1.5.3
npm install  # expect failure: Error: Module did not self-register.
rm vendor/<os-arch>/binding.node
npm install nan@1.5.3  # force nan 1.5.3 to be installed
npm install # Building

npm install # Says: Binary is fine; exiting
```

**Comment by am11** [comment]
@Xinkai, `nan` is already updated to latest v1.6.1: https://github.com/sass/node-sass/commit/196464492b6fcbda8f161c428ce35b786b9d8197.

**Comment by am11** [comment]
So the issue is with v1.6.1. Got it. :)

**Comment by Globegitter** [comment]
@am11 Yep there seems to be some issue with 1.6.1. 

Just updated a couple of forks and tried to test node-sass in a production [ember-cli](https://github.com/ember-cli/ember-cli) app. Getting it to npm install and build but running into some runtime errors :/
Might be a bit premature but if of any interest see: https://gist.github.com/Globegitter/4551c8860ad04e6d7618

**Comment by kkoopa** [comment]
1.6.2 is out

**Comment by am11** [comment]
@kkoopa, thanks! :+1:

**Comment by am11** [comment]
@kkoopa, tested on windows and it worked. Had to make this change:

```powershell
# powershell
git clone https://github.com/sass/node-sass --recursive  ; cd node-sass
git submodule --recursive --init
# update nan to v1.6.2 in package.json
npm install
iojs node_modules/pangyp/bin/node-gyp configure

## >>>> START_HACK

notepad build/binding.vcxproj
# Ctrl+H and replace node.lib to iojs.lib
# save and close

## >>>>  END_HACK

iojs

# this is iojs interactive console
> require('./').render({data: 'aside{margin: nolla}'})
# returns:
{ stats:
  { entry: 'data',
    start: 1423260352409,
    includedFiles: [],
    end: 1423260352409,
    duration: 3 },
  css: 'aside {\n margin: nolla; }\n',
  sourceMap: '{}' }
>
```

Would be nice if we don't have to apply that hack.. :)

**Comment by kkoopa** [comment]
It's a bug in pangyp

**Comment by ai** [comment]
nod-sass has same problem on `node.js 0.12`

**Comment by am11** [comment]
pangyp v2.0.2 is published with the fix. I have updated the package.json: 01ca5e2 via #652.

For `io.js`, you can now use:

```powershell
iojs node_modules/pangyp/bin/node-gyp rebuild
```

and for node:

```powershell
node node_modules/pangyp/bin/node-gyp rebuild
```

after `npm install`.

**Comment by Globegitter** [comment]
@am11 Awesome did get that to work now. But how can I install it as `npm` package in a project? Is there still much stopping that?

**Comment by am11** [comment]
@Globegitter, the upcoming node-sass v2.0.0 (stable) will support the present production-ready node.js version 0.12. Note that here *support* means the binaries available through dist. Otherwise you can manually build it with other versions of node.js et al.

**Comment by Globegitter** [comment]
@am11 Thanks. It is strange, so I have gotten the latest master to work in a fresh project, but if I reproduce the steps in an existing project I do not seem to be able to get it to work. So all builds fine, but then when I start my server I get the following error:

```js
dlopen(client/node_modules/broccoli-sass/node_modules/node-sass/vendor/darwin-x64/binding.node, 1): Symbol not found: __ZTVN4Sass6CssizeE
  Referenced from: client/node_modules/broccoli-sass/node_modules/node-sass/vendor/darwin-x64/binding.node
  Expected in: flat namespace
 in client/node_modules/broccoli-sass/node_modules/node-sass/vendor/darwin-x64/binding.node
Error: dlopen(client/node_modules/broccoli-sass/node_modules/node-sass/vendor/darwin-x64/binding.node, 1): Symbol not found: __ZTVN4Sass6CssizeE
  Referenced from: client/node_modules/broccoli-sass/node_modules/node-sass/vendor/darwin-x64/binding.node
  Expected in: flat namespace
 in client/node_modules/broccoli-sass/node_modules/node-sass/vendor/darwin-x64/binding.node
    at Error (native)
    at Module.load (module.js:341:32)
    at Function.Module._load (module.js:296:12)
    at Module.require (module.js:351:17)
    at require (module.js:370:17)
    at Object.<anonymous> (/client/node_modules/broccoli-sass/node_modules/node-sass/lib/index.js:192:15)
    at Module._compile (module.js:446:26)
    at Object.Module._extensions..js (module.js:464:10)
    at Module.load (module.js:341:32)
    at Function.Module._load (module.js:296:12)
```

Any idea what this might be about?

**Comment by am11** [comment]
@Globegitter, thanks for confirming that the `darwin` binaries are functional.

This feature request is completed. node-sass is compatible with `io.js`. You can build the binaries with the aforementioned steps. However the binaries distribution for io.js is a separate case, which we are tackling here: https://github.com/sass/node-sass/issues/655.

**Comment by am11** [comment]
Hey guys. node-sass v2.0.0 (superseded by patch release v2.0.1) is released with io.js v1.2 support. See the [release notes](https://github.com/sass/node-sass/releases/tag/v2.0.0) and the binary coverage matrix in [node-sass-binaries release notes](https://github.com/sass/node-sass-binaries/releases/tag/v2.0.0).

**Comment by mgenev** [comment]
now it broke on iojs 1.3... :-/

**Comment by xzyfer** [comment]
Libraries with native extension like libsass need to be compiled against
new versions of runtimes eg. node or io.js.

Technically it's not broken in 1.3 because 1.3 was never supported. If you
need support for 1.3 (or any future version) please open an issue
requesting it and we'll get to it as soon as we can.

For now downgrading to 1.2 or using nvm to run 1.2 in parallel will work.

**Comment by mgenev** [comment]
Hmm this will be a major chore in io's case as it seems to be releasing a new version every week....

**Comment by xzyfer** [comment]
I agree. I've opened a new issue to address the rapid release cycle of
io.js. _can't link on phone_

**Comment by Globegitter** [comment]
Just to have the cross-link: https://github.com/sass/node-sass/issues/694
